plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,228 @@
+#include <fc/log/logger_config.hpp> // set_os_thread_name
+
+#include <boost/filesystem.hpp> // C++17's filesystem not in our toolchain, use Boost's
+
+#include <thread>
+
+#include <sys/stat.h>
+
+namespace bfs = boost::filesystem;
+
+namespace eosio {
+   static appbase::abstract_plugin& _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
+
+class resource_monitor_plugin_impl {
+public:
+
+   void set_program_options(options_description&, options_description& cfg) {
+      cfg.add_options()
+         ( "resource-monitor-directory",  bpo::value<vector<bfs::path>>()->composing(),
+           "The directory whose file system to be monitored. This is a mandatory option. It can appear multiple times for different directories")
+         ( "resource-monitor-interval-seconds", bpo::value<uint32_t>()->default_value(2),
+           "Time in seconds between two consecutive checks of resource usage. Should be greater than 1 and less than 60 (one minute)" )
+         ( "resource-monitor-space-threshold", bpo::value<uint32_t>()->default_value(95),
+             "Threshold in terms of percentage of used space vs total space. If used space is over the threshold, a graceful shutdown is initiated. The value should be greater than 1 to less than 99" )
+         ;
+   }
+
+   void plugin_initialize(const appbase::variables_map& options) {
+      ilog("plugin_initialize");
+
+      // We must have at least one directory to monitor 
+      EOS_ASSERT(options.count("resource-monitor-directory") > 0, chain::plugin_config_exception, "At least one resource-monitor-directory must be configured");
+      const std::vector<bfs::path> dirs = options["resource-monitor-directory"].as<std::vector<bfs::path>>();
+
+      auto interval = options.at("resource-monitor-interval-seconds").as<uint32_t>();
+      EOS_ASSERT(interval > 0 && interval < 60, chain::plugin_config_exception,
+         "\"resource-monitor-interval-seconds\" must be greater than 0 and less than 60 (1 minute)");
+      sleep_time_in_ms = interval * 1000;  // Convert to milliseconds
+      ilog("interval = ${interval}", ("interval", interval));
+
+      used_space_threshold_in_percentage = options.at("resource-monitor-space-threshold").as<uint32_t>();
+      EOS_ASSERT(used_space_threshold_in_percentage > 0 && used_space_threshold_in_percentage < 100, chain::plugin_config_exception,
+         "\"resource-monitor-space-threshold\" must be greater than 0 and less than 100");
+      ilog("threshold = ${threshold}", ("threshold", used_space_threshold_in_percentage));
+
+      // Add filesystems containing the directories to the monitored list
+      // and set up actual threshold
+      for ( auto& dir: dirs ) {
+         add_file_system( dir );
+      }
+   }
+
+   // Start main thread
+   void plugin_startup() {
+      ilog("Creating and starting monitor thread");
+      monitor_thread = std::thread( [this] {
+         fc::set_os_thread_name( "resource_monitor" );
```

Comment:
`console_appender` uses 9 chars for thread name reporting. Recommend "resmon".

Reply:
Thanks Kevin for pointing out this! I will change the name to "resmon".

Reply:
Thanks Matthew for your comments. Somehow there is no reply under your comment. I was struggling when deciding the default value. Maybe 90% is more reasonable.